File: MATLAB/Ideal Dam Break/db_ERiemann.m
%   Solver Exato do problema de Riemann
% 
%   Solucao exata do problema de Riemann por amostragem, baseado na velocidade
%   das ondas que compoe o problema (choque ou rarefacao).
%   
%   Os testes suportados por esse programa sao: Escoamentos subcritico, 
%   transcritico e leito seco.
tic
clear all, close all, clc

%variaveis globais
global g dx m nint tol tmax
global xmin xmax dam

%constantes
g = 9.8; tol = 1e-6; nint = 50;  

%malha
xmin = -30; xmax = 30; m = 600;

%dados do problema
hL   = 3; hR = 0; vL = 0; vR = 0;
dam  = 0;

tmax = 2;

U = RiemannSolver(hL,hR,vL,vR);

for i = 1:m+1
    x(i) = xmin+(i-1)*dx-dam;
    if U(1,i) == 0
        v(i) = 0;
    else
        v(i) = U(2,i)/U(1,i);
    end
end

output(x,U,v);

figure(1)
plot(x,U(1,:),'b')
xlim([xmin xmax]), ylim([0 max(U(1,:))+0.5])
xlabel('x (m)'), ylabel('h+zb (m)')
title([tmax]), grid on

figure(2)
plot(x,v,'b')
xlim([xmin xmax]), ylim([min(v)-0.5 max(v)+0.5])
xlabel('x (m)'), ylabel('v (m/s)')
title([tmax]), grid on 

figure(3)
plot(x,U(2,:),'b')
xlim([xmin xmax]), ylim([min(U(2,:))-0.1 max(U(2,:))+0.1])
xlabel('x (m)'), ylabel('q (m^2/s)')
title([tmax]), grid on 

toc

%-----------------------------------------------------------------------
%Escolha do solver de Riemann - Molhado(wet) ou Seco(dry)
%-----------------------------------------------------------------------
function U = RiemannSolver(hL,hR,vL,vR)
    global g
    
    cL = sqrt(g*hL); 
    cR = sqrt(g*hR);
    cs = (vR-vL)-2*(cL+cR);
    
    if hL == 0 | hR ==0 | cs >= 0
%       call dry solver
        U = drybed(hL,hR,vL,vR,cL,cR);
    else
%       call wet solver
        U = wetbed(hL,hR,vL,vR,cL,cR);
    end
end
%-----------------------------------------------------------------------
%Solver de Riemann para leito seco (drybed)
%-----------------------------------------------------------------------
function U = drybed(hL,hR,vL,vR,cL,cR)
    global g dx m tmax
    global xmin xmax dam
    
    U = zeros(2,m+1);
    
    dx = (xmax-xmin)/m;
    
    for i = 1:m+1
        x = xmin+(i-1)*dx-dam;
        s = x/tmax;
        
        if hL == 0
%           Leito seco a esquerda
            [h,v] = drylef(hL,hR,vL,vR,cL,cR,s);
        else
            if hR == 0
%               Leito seco a direita
                [h,v] = dryrig(hL,hR,vL,vR,cL,cR,s);
            else
%               Leito seco no meio
                [h,v] = drymid(hL,hR,vL,vR,cL,cR,s);
            end
        end
        U(1,i) = h;
        U(2,i) = h*v;
    end
end
%-----------------------------------------------------------------------
%Construcao da solucao para leito seco a esquerda (dry left, hL = 0)
%-----------------------------------------------------------------------
function [h,v] = drylef(hL,hR,vL,vR,cL,cR,s)
    global g
    
    sR = vR+cR;
    if s >= sR
        h = hR;
        v = vR;
    else
        sRs = vR-2*cR;
        if s >= sRs
            h = (1/3*(-vR+2*cR+s))^2/g;
            v = 1/3*(vR-2*cR+2*s);
        else
            h = hL;
            v = vL;
        end
    end
end
%-----------------------------------------------------------------------
%Construcao da solucao para leito seco a direita (dry right, hR = 0)
%-----------------------------------------------------------------------
function [h,v] = dryrig(hL,hR,vL,vR,cL,cR,s)
    global g
    
    sL = vL-cL;
    if s <= sL
        h = hL;
        v = vL;
    else
        sLs = vL+2*cL;
        if s <= sLs
            h = (1/3*(vL+2*cL-s))^2/g;
            v = 1/3*(vL+2*cL+2*s);
        else
            h = hR;
            v = vR;
        end
    end
end
%-----------------------------------------------------------------------
%Construcao da solucao para leito seco no meio (hs = 0)
%-----------------------------------------------------------------------
function [h,v] = drymid(s,hL,hR,vL,vR,cL,cR)
    global g
    
    sL = vL-cL; sLs = vL+2*cL;
    sR = vR+cR; sRs = vR-2*cR;
    
    if s <= sL
        h = hL;
        v = vL;
    end
    
    if s > sL & s <= sLs
        h = (1/3*(vL+2*cL-s))^2/g;
        v = 1/3*(vL+2*cL+2*s);
    end
    
    if s > sLs & s <= sRs
        h = 0;
        v = 0;
    end
    
    if s > sRs & s <= sR
        h = (1/3*(-vR+2*cR+s))^2/g;
        v = 1/3*(vR-2*cR+2*s);
    end
    
    if s > sR
        h = hR;
        v = vR;
    end
end
%-----------------------------------------------------------------------
%Solver de Riemann para leito molhado (wetbed)
%-----------------------------------------------------------------------
function U = wetbed(hL,hR,vL,vR,cL,cR)
    global g dx m nint tol tmax
    global xmin xmax dam
    
    U = zeros(2,m+1);
    
    dx = (xmax-xmin)/m;
    
%   Chute inicial para o metodo de Newton-Raphson    
    hs = 1/g*(0.5*(cL+cR)+0.25*(vL-vR))^2;
    hz = hs;
    
    for i = 1:nint
        cs = sqrt(g*hs);

%       Fluxo numerico para o metodo de Newton-Raphson        
        [fL,dfL] = Nflux(hL,hs,vL,cL,cs);
        [fR,dfR] = Nflux(hR,hs,vR,cR,cs);
        
        f = fR+fL+vR-vL;
        df = dfL+dfR;
        hs = hs-f/df;
        
        if abs(hs-hz) <= tol
            break
        else
            hz = hs;
        end
    end
    
    cs = sqrt(g*hs);
    vs = 0.5*(vR+vL+fR-fL);
    
%   Solucao por amostragem    
    for i = 1:m+1
        x = xmin+(i-1)*dx-dam;
        s = x/tmax;
        
        [h,v] = samwet(hL,hR,hs,vL,vR,vs,cL,cR,cs,s);
        
        U(1,i) = h;
        U(2,i) = h*v;
    end
end
%-----------------------------------------------------------------------
%Avaliacao da funcao f(h*) pelo metodo de Newton
%-----------------------------------------------------------------------
function [fK,dfK] = Nflux(hK,hs,vK,cK,cs)
    global g 
    
%   Onda K(Left or Right)
    if hs <= hK
        fK = 2*(cs-cK);
        dfK = g/cs;
    else
        fK = (hs-hK)*sqrt((hs+hK)*g/(2*hs*hK));
        dfK = sqrt((hs+hK)*g/(2*hs*hK))...
            -(hs-hK)*g/(4*hs^2)*1/sqrt((hs+hK)*g/(2*hs*hK));
    end
end
%-----------------------------------------------------------------------
%Construcao da solucao para leito molhado
%-----------------------------------------------------------------------
function [h,v] = samwet(hL,hR,hs,vL,vR,vs,cL,cR,cs,s)
    global g
    
    if s <= vs
%       Onda a esquerda
        if hL <= hs
%           Onda de choque
            lbL = sqrt(0.5*(hs+hL)*hs/hL^2);
            sL = vL-lbL*cL;
            if s <= sL
                h = hL;
                v = vL;
            else
                h = hs;
                v = vs;
            end
        else
%           Onda de rarefacao
            sL = vL-cL;
            if s <= sL
                h = hL;
                v = vL;
            else
                sLs = vs-cs;
                if s <= sLs
                    h = (1/3*(vL+2*cL-s))^2/g;
                    v = 1/3*(vL+2*cL+2*s);
                else
                    h = hs;
                    v = vs;
                end
            end
        end
    else
%       Onda a direita
        if hR <= hs
%           Onda de choque
            lbR = sqrt(0.5*(hs+hR)*hs/hR^2);
            sR = vR+lbR*cR;
            if s >= sR
                h = hR;
                v = vR;
            else
                h = hs;
                v = vs;
            end
        else
%           Onda de rarefacao
            sR = vR+cR;
            if s >= sR
                h = hR;
                v = vR;
            else
                sRs = vs+cs;
                if s >= sRs
                    h = (1/3*(-vR+2*cR+s))^2/g;
                    v = 1/3*(vR-2*cR+2*s);
                else
                    h = hs; 
                    v = vs;
                end
            end
        end
    end
end
%-----------------------------------------------------------------------
%Geracao do arquivo de saida
%-----------------------------------------------------------------------
function output(x,U,v)
    outfile = [x;U;v];
    fileID = fopen('db_ERiemann.out','w');
    fprintf(fileID,'%6.2f %12.8f %12.8f %12.8f\r\n',outfile);
    fclose(fileID);
end

File: Fortran/Ideal Dam Break/db_ERiemann.f90
!   Solver Exato do problema de Riemann 
! 
!   Solução exata do problema de Riemann por amostragem, baseado na velocidade
!   das ondas que compõe o problema (choque ou rarefação)
! 
!   Os testes suportados por esse programa são: Escoamentos subcrítico, 
!   transcrítico e leito seco.

program DB_ERiemann
    implicit none
    
    real :: tic, toc
    
    call cpu_time(tic)
    
!   Leitura do arquivo .ini que contém as condições iniciais 
    call leitura

!   Determinação do tipo do problema e construção da solução
    call RiemannSolver

!   Escrita das solução obtida     
    call output
    
!   Executa o plot da solução 
    call execute_command_line('gnuplot -p plot_db_ERiemann.plt . &>/dev/null')
    
    call cpu_time(toc)
            
    write(*,*) '--------------------------------'           
    write(*,*) 'Tempo de execução   = ', toc-tic, 'segundos'
    write(*,*) '--------------------------------'    

end program DB_ERiemann
!--------------------------------------------------------------------
!Leitura dos dados de entrada
!--------------------------------------------------------------------
subroutine leitura
    implicit none
    
    double precision     :: g, xmin, xmax, dx, tol, dam
    double precision     :: hL, hR, vL, vR, cL, cR, tmax 
    integer :: m, nint
    
    common /cini/ hL, hR, vL, vR, cL, cR
    common /malh/ xmin, xmax, dx, dam, m
    common /tole/ tol, nint
    common /temp/ tmax
    common /grav/ g
    
    open(unit = 1, file = 'db_ERiemann.ini', status = 'unknown')
    
    read(1,*) xmin      ! Ponto inicial do canal
    read(1,*) xmax      ! Ponto final do canal 
    read(1,*) dam       ! Posição da barragem
    read(1,*) hL        ! Altura da lamina d'água à esquerda da descontinuidade   
    read(1,*) hR        ! Altura da lamina d'água à direita da descontinuidade       
    read(1,*) vL        ! Velocidade à esquerda da descontinuidade
    read(1,*) vR        ! Velocidade à direita da descontinuidade
    read(1,*) tmax      ! Instante do tempo para solução final
    read(1,*) tol       ! Tolerancia para o método de Newton
    read(1,*) nint      ! Numero maximo de iterações no método de Newton
    read(1,*) m         ! Numero de pontos/celulas no dominio
    read(1,*) g         ! Aceleração da gravidade
    
    close(1)
    
    write(*,*) '--------------------------------'
    write(*,*) 'Dados iniciais:'
    write(*,*) '--------------------------------'
    write(*,*) 'xmin  = ', xmin
    write(*,*) 'xmax  = ', xmax
    write(*,*) 'dam   = ', dam
    write(*,*) 'hL    = ', hL
    write(*,*) 'hR    = ', hR
    write(*,*) 'vL    = ', vL
    write(*,*) 'vR    = ', vR
    write(*,*) 'tmax  = ', tmax
    write(*,*) 'tol   = ', tol
    write(*,*) 'nint  = ', nint
    write(*,*) 'm     = ', m
    write(*,*) 'g     = ', g
    write(*,*) '--------------------------------'
    
end subroutine leitura
!--------------------------------------------------------------------
!Escolha do solver de Riemann
!--------------------------------------------------------------------
subroutine RiemannSolver
    implicit none

    double precision :: g, hL, hR, hs, vL, vR, vs, cL, cR, cs
       
    common /cini/ hL, hR, vL, vR, cL, cR
    common /star/ hs, vs, cs
    common /grav/ g
    
    cL = sqrt(g*hL)
    cR = sqrt(g*hR)
    cs = (vR-vL)-2*(cL+cR)
    
    if (hL.EQ.0.OR.hR.EQ.0.OR.cs.GE.0) then 
        call drybed
    else
        call wetbed
    end if
    
end subroutine RiemannSolver
!--------------------------------------------------------------------
!Solver de Riemann para leito seco (drybed)
!--------------------------------------------------------------------
subroutine drybed
    implicit none
    
    double precision :: g, xmin, xmax, dx, tmax, dam, x, h, v, s
    double precision :: hL, hR, vL, vR, cL, cR
    integer :: i, m 
    integer, parameter :: mx = 5000 
    double precision, dimension (2,mx+1) :: U
    
    common /cini/ hL, hR, vL, vR, cL, cR
    common /malh/ xmin, xmax, dx, dam, m
    common /temp/ tmax
    common /grav/ g
    common /vari/ U
    
    dx = (xmax-xmin)/dble(m)

!   Construção da solução      
    do i = 1,m+1
        x = xmin+(i-1)*dx-dam
        s = x/tmax
        
        if (hL.EQ.0) then
            call drylef(h,v,s)
        else
            if (hR.EQ.0) then
                call dryrig(h,v,s)
            else
                call drymid(h,v,s)
            end if
        end if
        
        U(1,i) = h
        U(2,i) = h*v
    end do
        
end subroutine drybed
!--------------------------------------------------------------------
!Construção da solução para o caso de leito seco à esquerda (dry left, hL = 0)
!--------------------------------------------------------------------
subroutine drylef(h,v,s)
    implicit none
    
    double precision :: g, h, v, s, sR, sRs
    double precision :: hL, hR, vL, vR, cL, cR
    
    common /cini/ hL, hR, vL, vR, cL, cR
    common /grav/ g
    
    sR = vR+cR
    if (s.GE.sR) then
        h = hR
        v = vR
    else
        sRs = vR-2*cR
        if (s.GE.sRs) then
            h = ((-vR+2*cR+s)**2)/(9*g)
            v = (vR-2*cR+2*s)/3
        else
            h = hL
            v = vL
        end if
    end if
    
end subroutine drylef
!--------------------------------------------------------------------
!Construção da solução para o caso de leito seco à direita (dry right, hR = 0)
!--------------------------------------------------------------------
subroutine dryrig(h,v,s)
    implicit none
    
    double precision :: g, h, v, s, sL, sLs
    double precision :: hL, hR, vL, vR, cL, cR
    
    common /cini/ hL, hR, vL, vR, cL, cR
    common /grav/ g
    
    sL = vL-cL
    if (s.LE.sL) then
        h = hL
        v = vL
    else
        sLs = vL+2*cL
        if (s.LE.sLs) then
            h = (vL+2*cL-s)**2/(9*g)
            v = (vL+2*cL+2*s)/3
        else
            h = hR
            v = vR
        end if
    end if
    
end subroutine dryrig
!--------------------------------------------------------------------
!Construção da solução para o caso de leito seco no meio (hs = 0)
!--------------------------------------------------------------------
subroutine drymid(h,v,s)
    implicit none
    double precision :: g, h, v, s, sL, sR, sLs, sRs
    double precision :: hL, hR, vL, vR, cL, cR
    
    common /cini/ hL, hR, vL, vR, cL, cR
    common /grav/ g
    
    sL = vL-cL 
    sR = vR+cR 
    
    sLs = vL+2*cL
    sRs = vR-2*cR

    if (s.LE.sL) then
        h = hL
        v = vL
    end if
        
    if (s.GT.sL.AND.s.LE.sLs) then
        h = (vL+2*cL-s)**2/(9*g)
        v = (vL+2*cL+2*s)/3
    end if
        
    if (s.GT.sLs.AND.s.LE.sRs) then
        h = 0
        v = 0
    end if
        
    if (s.GT.sRs.AND.s.LE.sR) then
        h = (-vR+2*cR+s)**2/(9*g)
        v = (vR-2*cR+2*s)/3
    end if
        
    if (s.GT.sR) then
        h = hR
        v = vR
    end if

end subroutine drymid
!--------------------------------------------------------------------
!Solver de Riemann para leito molhado (wetbed)
!--------------------------------------------------------------------
subroutine wetbed
    implicit none

    double precision :: g, xmin, xmax, dx, tmax, dam, tol, x, h, v, s
    double precision :: f, df, hz, fL, fR, dfL, dfR
    double precision :: hL, hR, hs, vL, vR, vs, cL, cR, cs
    integer :: i, k, m, nint 
    integer, parameter :: mx = 5000 
    double precision, dimension (2,mx+1) :: U
    
    common /cini/ hL, hR, vL, vR, cL, cR
    common /malh/ xmin, xmax, dx, dam, m
    common /star/ hs, vs, cs
    common /tole/ tol, nint
    common /temp/ tmax
    common /grav/ g
    common /vari/ U
    
    dx = (xmax-xmin)/dble(m)
    
    hs = 1/g*(0.5*(cL+cR)+0.25*(vL-vR))**2
    hz = hs
    
!   Aplicação do método de Newton para determinar as variaveis na região estrela    
    do k = 1,nint
    
        cs = sqrt(g*hs)
        
        call Nflux(fL,dfL,hL)
        call Nflux(fR,dfR,hR)
        
        f  = fR+fL+vR-vL
        df = dfL+dfR
        
        hs = hs-f/df
        
        if (abs(hs-hz).LE.tol) then
            exit
        else
            hz = hs
        end if
    end do
    
    write(*,*) '--------------------------------'           
    write(*,*) 'Numero de iterações = ', k
    write(*,*) '--------------------------------'
    
    cs = sqrt(g*hs)
    vs = 0.5*(vR+vL+fR-fL)

!   Construção da solução    
    do i = 1,m+1
        x = xmin+(i-1)*dx-dam
        s = x/tmax

        call samwet(h,v,s)
        
        U(1,i) = h
        U(2,i) = h*v
    end do
    
end subroutine wetbed
!--------------------------------------------------------------------
!Avaliação da função f(h*) pelo método de Newton
!--------------------------------------------------------------------
subroutine Nflux(fK,dfK,hK)
    implicit none
    
    double precision :: g, fK, dfK, hK, cK
    double precision :: hs, vs, cs
    
    common /star/ hs, vs, cs
    common /grav/ g
    
    cK = sqrt(g*hK)
    
    if (hs.LE.hK) then
        fK = 2*(cs-cK)
        dfK = g/cs
    else
        fK = (hs-hK)*sqrt((hs+hK)*g/(2*hs*hK))
        dfK = sqrt((hs+hK)*g/(2*hs*hK))-(hs-hK)*g/(4*hs**2)*1/sqrt((hs+hK)*g/(2*hs*hK))
    end if 
    
end subroutine Nflux
!--------------------------------------------------------------------
!Construção da solução para o caso de leito molhado
!--------------------------------------------------------------------
subroutine samwet(h,v,s)
    implicit none
    
    double precision :: g, h, v, s, lbL, lbR, sL, sLs, sR, sRs
    double precision :: hL, hR, hs, vL, vR, vs, cL, cR, cs
    
    common /cini/ hL, hR, vL, vR, cL, cR
    common /star/ hs, vs, cs
    common /grav/ g

    if (s.LE.vs) then
!       Onda á esquerda
        if (hL.LE.hs) then
!           Onda de choque
            lbL = sqrt(0.5*(((hs+hL)*hs)/hL**2))
            sL = vL-lbL*cL
            if (s.LE.sL) then 
                h = hL
                v = vL
            else
                h = hs
                v = vs
            end if
        else
!           Onda de rarefação
            sL = vL-cL
            if (s.LE.sL) then 
                h = hL
                v = vL
            else
                sLs = vs-cs
                if (s.LE.sLs) then
                    h = ((vL+2*cL-s)**2)/(9*g)
                    v = (vL+2*cL+2*s)/3
                else 
                    h = hs
                    v = vs
                end if
            end if
        end if    
    else
!       Onda á direita
        if (hR.LE.hs) then
!           Onda de choque        
            lbR =  sqrt(0.5*(((hs+hR)*hs)/hR**2))
            sR = vR+lbR*cR
            if (s.GE.sR) then 
                h = hR
                v = vR
            else
                h = hs
                v = vs
            end if
        else
!           Onde de rarefação        
            sR = vR+cR
            if (s.GE.sR) then 
                h = hR
                v = vR
            else
                sRs = vs+cs
                if (s.GE.sRs) then
                    h = ((-vR+2*cR+s)**2)/(9*g)
                    v = (vR-2*cR+2*s)/3
                else 
                    h = hs
                    v = vs
                end if
            end if
        end if
    end if
end subroutine samwet
!--------------------------------------------------------------------
!Geração do arquivo de saída
!--------------------------------------------------------------------
subroutine output
    implicit none
    
    double precision :: xmin, xmax, dx, dam, x
    integer :: i, m
    integer, parameter :: mx = 5000 
    double precision, dimension (2,mx+1) :: U
    
    common /malh/ xmin, xmax, dx, dam, m
    common /vari/ U
    
    dx = (xmax-xmin)/dble(m)
    
    open(unit = 1, file = 'db_ERiemann.out', status = 'unknown')
    do i = 1,m+1
        x = xmin+(i-1)*dx-dam
        
        if (U(1,i).GT.0) then        
            write(1,*) x, U(1,i), U(2,i), U(2,i)/U(1,i)
        else
            write(1,*) x, U(1,i), U(2,i), 0
        end if
    end do

    close(1)
     
end subroutine output
